### ISSUE TITLE
Some controller endpoints fail when trying to serialize output.

### ISSUE THREAD

**Issue by jMyles** [nucypher/nucypher#920]
This is an insidious issue.  

We take string labels for our endpoints (as JSON).  These strings get encoded as b64.  Then, when the response is parsed, they are rendered back to UTF-8 for display as JSON.

This means that 1 out of every 8 labels is invalid.  For example, `'516d593559505355376d454b61374751577146467a47754658396d516a685674716b7663744b376b4b666a35336d'` is not valid, because its b64 representation can't be serialized as UTF-8.

Here's a discussion of this matter: https://ptb.discordapp.com/channels/411401661714792449/411401661714792451/564353305887637517

@KPrasch suggests in #508 that we adopt UTF-8; I can now fully concur.

**Comment by cygnusv** [comment]
Wait, are you suggesting to use UTF8 to encode arbitrary bytestrings? That's a [no-no](https://en.wikipedia.org/wiki/UTF-8#Invalid_byte_sequences). What's exactly your suggestion then?

Another thing:
>  For example, '516d593559505355376d454b61374751577146467a47754658396d516a685674716b7663744b376b4b666a35336d' is not valid, because its b64 representation can't be serialized as UTF-8.

Can you clarify this more? I don't understand how the output of a base64 encoding, which is always an ASCII string cannot be serialized as UTF-8. There may be something I'm not understanding right. I also tried to do what you describe and I don't have any issues:

```python
>>> x = b'516d593559505355376d454b61374751577146467a47754658396d516a685674716b7663744b376b4b666a35336d'
>>> y = base64.b64encode(x)
>>> y
b'NTE2ZDU5MzU1OTUwNTM1NTM3NmQ0NTRiNjEzNzQ3NTE1NzcxNDY0NjdhNDc3NTQ2NTgzOTZkNTE2YTY4NTY3NDcxNmI3NjYzNzQ0YjM3NmI0YjY2NmEzNTMzNmQ='
>>> z = y.decode("utf-8")
>>> z
'NTE2ZDU5MzU1OTUwNTM1NTM3NmQ0NTRiNjEzNzQ3NTE1NzcxNDY0NjdhNDc3NTQ2NTgzOTZkNTE2YTY4NTY3NDcxNmI3NjYzNzQ0YjM3NmI0YjY2NmEzNTMzNmQ='
>>> w = base64.b64decode(z)
>>> w
b'516d593559505355376d454b61374751577146467a47754658396d516a685674716b7663744b376b4b666a35336d'
>>> x == w
True
```
It also works when the input is not a bytestring, but a string:
```python
x = '516d593559505355376d454b61374751577146467a47754658396d516a685674716b7663744b376b4b666a35336d'
>>> y = base64.b64encode(x.encode("utf-8"))
```

**Comment by jMyles** [comment]
> Wait, are you suggesting to use UTF8 to encode arbitrary bytestrings?

Of course not.

> What's exactly your suggestion then?

To require that labels be UTF-8 compliant bytestrings or (in the case of the python API) native unicode strings.

> Can you clarify this more? I don't understand how the output of a base64 encoding, which is always an ASCII string cannot be serialized as UTF-8.

Surely you meant *deserialized*?

This is a non-sequitur.  I can represent an ASCII string with smoke signals too - that doesn't make UTF-8 a decoder chimney.  Of course there are ways to encode a string that cannot subsequently be represented as unicode via UTF-8 decoding. 

```
>>> x = '516d593559505355376d454b61374751577146467a47754658396d516a685674716b7663744b376b4b666a35336d'
>>> y = base64.b64encode(x.encode("utf-8"))
```
Huh?  What does that have to do with the price of tea in China?

I think you probably meant this (which, as you can see, causes the error in question):

```
In [1]: import base64                                                           

In [2]: x = '516d593559505355376d454b61374751577146467a47754658396d516a685674716
   ...: b7663744b376b4b666a35336d'                                              

In [3]: y = base64.b64decode(x)                                                 

In [4]: y.decode("utf-8")                                                       
---------------------------------------------------------------------------
UnicodeDecodeError                        Traceback (most recent call last)
<ipython-input-4-d3ff2b271d40> in <module>
----> 1 y.decode("utf-8")

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe7 in position 0: invalid continuation byte
```

**Comment by cygnusv** [comment]
> > What's exactly your suggestion then?
> 
> To require that labels be UTF-8 compliant bytestrings or (in the case of the python API) native unicode objects.

OK, yeah, that makes perfect sense.

> Have you taken a look at the test in #921? It clearly shows (one manifestation of) the problem that is blocking @robin-thomas. Is there any doubt in your mind, at this time, that it's possible to cause this error simply by passing an incompatible label to a controller endpoint?

I have seen the test case, but it's too high-level for me to clearly understand the underlying problem. I needed something more low-level, which is what you provided in your response above (thanks for that, btw!).

I think what was confusing to me was that you said that "the _b64 representation of '516....36d' can't be serialized as UTF-8_". At least for me, the b64 representation of some `x` is `base64.b64encode(x)`. It seems that you were talking about the opposite, `base64.b64decode(x)`.

**Comment by jMyles** [comment]
> I think what was confusing to me was that you said that "the b64 representation of '516....36d' can't be serialized as UTF-8".

Ahh, I understand.  Yes, this is a misstatement. What I meant is:

> Some unicode strings - proper labels if we adopt unicode as our label format - are decoded into base64 using bytes that can't be represented via UTF-8 (and thus, their bytes representation can't be represented as JSON with our current approach).

**Comment by jMyles** [comment]
Is this issue closed?

I think that, pursuant to our discussions on Discord (starting [here](https://ptb.discordapp.com/channels/411401661714792449/411401661714792451/587144270318927887) but continuing only intermittently), it's fair to say that @tuxxy, @michwill, and @cygnusv are of a mindset that's in the ballpark of "this issue is fixed by `62c7df293a03347b16967fe7de7924199c38f819` and similar patterns."

Put another way, these bytes-hungry colleagues of mine are content to understand that anybody who calls `grant` (currently, if we're taking attendance, that's 1) the CLI, 2) the HTTP endpoints, 3) the IPC / unix tooling, our users who awesomely choose to use the Python API directly) is supposed to make a design decision about how to turn unicode characters into bytes.

In the 3 cases over which we have control, we have chosen UTF-8 (although @tuxxy correctly points out that UTF-16 takes an even broader view of lingual inclusion, albeit at a huge memory cost).

My preference, although I don't feel terribly strongly about it, is that `grant` itself is better understood as the conjoining place where unicode is accepted - in other words, that the proper type hint for `label` in `grant(...)` is `str`.

This is chiefly to avoid a scenario wherein the person sharing data does, eg, the following:

`nucypher alice grant ... ❽◯ ▲ ▶ஔ`

...and then shares that label, in unicode, with a would-be recipient, who is using an application developed for the same ecosystem, but which is similarly agnostic regarding encoding as @tuxxy and @michwill and @cygnusv want to be.

Now, they are left guessing - which encoding are they supposed to use?  UTF-8 is of course the correct answer, but how are they to know that?  Well, they have to check our documentation of course.

And if we're *forcing* them to use UTF-8 (ie, we have decided that our own interfaces are gnostic on this matter) then why don't we simply make this a default?

Put another way, if you reeeaaaalllly want to make a policy with a UTF-8 non-compliant label, then you can do that via the `create_policy` Python API.  Otherwise, we assume that you're in a unicode zone.

But, if that's actually the design we want (people to need to guess about what encoding to use in the Python API to have compatibility with the other interfaces, but also the flexibility to use whatever bytes they want) then I guess I can live with that.  I just want to hear everybody acknowledge that that's the price here.
